Nodejs/Product/Npm/IPackageCatalog.cs

```diff
@@ -16,16 +16,146 @@
 
 using System;
 using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
 using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
 
 namespace Microsoft.NodejsTools.Npm {
     public interface IPackageCatalog {
         DateTime LastRefreshed { get; }
 
-        Task<IEnumerable<IPackage>> GetCatalogPackagesAsync(string filterText, Uri registryUrl = null);
-
-        IPackage this[string name] { get; }
+        Task<IEnumerable<IPackage>> GetCatalogPackagesAsync(string filterText);
 
         long? ResultsCount { get; }
     }
+
+    // This class is a wrapper catalog to directly query the NPM repo
+    // instead of downloading the entire catalog (which is no longer supported).
+    internal sealed class EmptyPackageCatalog : IPackageCatalog {
+
+        private static readonly Uri defaultRegistryUri = new Uri("https://registry.npmjs.org/");
+
+        public static readonly IPackageCatalog Instance = new EmptyPackageCatalog();
+
+        private EmptyPackageCatalog() { }
+
+        public DateTime LastRefreshed => DateTime.Now;
+
+        public long? ResultsCount => 0;
+
+        public async Task<IEnumerable<IPackage>> GetCatalogPackagesAsync(string filterText) {
+            var relativeUri = string.Format("/-/v1/search?text={0}", filterText);
+            var searchUri = new Uri(defaultRegistryUri, relativeUri);
+
+            var request = WebRequest.Create(searchUri);
+            using (var response = await request.GetResponseAsync()) {
+                var reader = new StreamReader(response.GetResponseStream());
+                using (var jsonReader = new JsonTextReader(reader)) {
+                    while (jsonReader.Read()) {
+                        switch (jsonReader.TokenType) {
+                            case JsonToken.StartObject:
+                            case JsonToken.PropertyName:
+                                continue;
+                            case JsonToken.StartArray:
+                                return ReadPackagesFromArray(jsonReader);
+                            default:
+                                throw new InvalidOperationException("Unexpected json token.");
+                        }
+                    }
+                }
+            }
+
+            // should never get here
+            throw new InvalidOperationException("Unexpected json token.");
```

Comment:
What do these throws cause to happen (not just this, but in a number of places in this code if the data isn't as expected)? We shouldn't bring down VS if some remote API returns incorrect data. Hopefully they just get swallowed further up the chain. Can you verify and add a comment?